File: Algorithms/GPU/histogram_equalization/openACC/equalization.cpp
// 






#include "opencv2/core/core.hpp"
#include "opencv2/imgproc/imgproc.hpp"
#include "opencv2/highgui/highgui.hpp"
#include <iostream>
#include <cmath>
#include <string>
#include "nvToolsExt.h"

using namespace std;
using namespace cv;

int main() {

    Mat image = imread("./input/trees_big.jpg", IMREAD_GRAYSCALE);
    if (image.empty()) {
        cerr << "Failed to load ./input/trees_big.jpg" << endl;
        return -1;
    }
    Mat myMat1;

    uchar* img_data = image.ptr<uchar>(0);
    int rows = image.rows;
    int cols = image.cols;
    int size = rows * cols;


    for (int iter = 0; iter < 15; iter++) {

        string iter_name = "iteration_" + to_string(iter);
        nvtxRangePushA(iter_name.c_str());

        int arr[256] = {0};
        float arr2[256] = {0};
        float arr3[256] = {0};
        int h2[256] = {0};

        // ------------------------ histogram_original ------------------------
        nvtxRangePushA("histogram_original");

        #pragma acc enter data copyin(arr[0:256], img_data[0:size])

        #pragma acc parallel loop collapse(2) present(img_data[0:size], arr[0:256])
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                int index = static_cast<int>(img_data[i * cols + j]);
                #pragma acc atomic
                arr[index]++;
            }
        }

        #pragma acc exit data copyout(arr[0:256])

        nvtxRangePop();

        // ------------------------ find_max_original_hist --------------------
        nvtxRangePushA("find_max_original_hist");

        int maxVal = 0;
        for (int i = 0; i < 256; i++) {
            if (arr[i] > maxVal) {
                maxVal = arr[i];
            }
        }

        nvtxRangePop();

        // ------------------------ pmf_compute -------------------------------
        nvtxRangePushA("pmf_compute");

        float total = static_cast<float>(size);
        for (int i = 0; i < 256; i++) {
            arr2[i] = static_cast<float>(arr[i]) / total;
        }

        nvtxRangePop();

        // ------------------------ cdf_compute -------------------------------
        nvtxRangePushA("cdf_compute");

        arr3[0] = arr2[0];
        for (int i = 1; i < 256; i++) {
            arr3[i] = arr2[i] + arr3[i - 1];
        }

        nvtxRangePop();

        // ------------------------ apply_equalization -----------------------
        nvtxRangePushA("apply_equalization");

        myMat1 = Mat(rows, cols, CV_8U, Scalar(0));
        uchar* out_data = myMat1.ptr<uchar>(0);

        #pragma acc enter data copyin(arr3[0:256]) create(out_data[0:size])

        #pragma acc parallel loop collapse(2) present(img_data[0:size], out_data[0:size], arr3[0:256])
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                int idx = i * cols + j;
                out_data[idx] = static_cast<uchar>(floor(255.0f * arr3[img_data[idx]]));
            }
        }

        #pragma acc exit data copyout(out_data[0:size]) delete(arr3[0:256])

        nvtxRangePop();

        // ------------------------ histogram_equalized -----------------------
        nvtxRangePushA("histogram_equalized");

        #pragma acc enter data copyin(out_data[0:size], h2[0:256])

        #pragma acc parallel loop collapse(2) present(out_data[0:size], h2[0:256])
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                int idx = i * cols + j;
                int index = static_cast<int>(out_data[idx]);
                #pragma acc atomic
                h2[index]++;
            }
        }

        #pragma acc exit data copyout(h2[0:256]) delete(out_data[0:size], img_data[0:size])

        nvtxRangePop();

        // ------------------------ find_max_equalized_hist -------------------
        nvtxRangePushA("find_max_equalized_hist");

        int maxH2 = 0;
        for (int i = 0; i < 256; i++) {
            if (h2[i] > maxH2) {
                maxH2 = h2[i];
            }
        }

        nvtxRangePop();

        nvtxRangePop(); 
    }

    imwrite("./output/equalized_acc.jpg", myMat1);
    cout << "Output saved to ./output/equalized_acc.jpg" << endl;

    return 0;
}

File: Algorithms/GPU/histogram_equalization/CUDA_Python/equalization.py
import numpy as np
import cv2
from numba import cuda
import math
import nvtx

# CUDA-specific: Shared memory size constant
HIST_BINS = 256

# ------------------------ CUDA Kernels ------------------------

# Histogram kernel using shared memory (CUDA-specific optimization)
@cuda.jit
def histogram_kernel(img_data, arr, rows, cols):
    # CUDA-specific: Shared memory allocation
    local_hist = cuda.shared.array(256, dtype=np.int32)
    
    # CUDA-specific: Thread indexing
    tid = cuda.threadIdx.y * cuda.blockDim.x + cuda.threadIdx.x
    i = cuda.blockIdx.y * cuda.blockDim.y + cuda.threadIdx.y
    j = cuda.blockIdx.x * cuda.blockDim.x + cuda.threadIdx.x
    
    # Initialize shared memory
    if tid < 256:
        local_hist[tid] = 0
    
    # CUDA-specific: Block-level synchronization
    cuda.syncthreads()
    
    # Compute histogram in shared memory (faster atomics)
    if i < rows and j < cols:
        index = int(img_data[i, j])
        # CUDA-specific: Shared memory atomic (fast)
        cuda.atomic.add(local_hist, index, 1)
    
    cuda.syncthreads()
    
    # Reduce shared histogram to global memory
    if tid < 256:
        # CUDA-specific: Global memory atomic (only 256 per block)
        cuda.atomic.add(arr, tid, local_hist[tid])


# Apply equalization kernel
@cuda.jit
def apply_equalization_kernel(img_data, out_data, arr3, rows, cols):
    i = cuda.blockIdx.y * cuda.blockDim.y + cuda.threadIdx.y
    j = cuda.blockIdx.x * cuda.blockDim.x + cuda.threadIdx.x
    
    if i < rows and j < cols:
        index = int(img_data[i, j])
        out_data[i, j] = int(math.floor(255.0 * arr3[index]))


# Histogram equalized kernel using shared memory
@cuda.jit
def histogram_equalized_kernel(out_data, h2, rows, cols):
    local_hist = cuda.shared.array(256, dtype=np.int32)
    
    tid = cuda.threadIdx.y * cuda.blockDim.x + cuda.threadIdx.x
    i = cuda.blockIdx.y * cuda.blockDim.y + cuda.threadIdx.y
    j = cuda.blockIdx.x * cuda.blockDim.x + cuda.threadIdx.x
    
    if tid < 256:
        local_hist[tid] = 0
    
    cuda.syncthreads()
    
    if i < rows and j < cols:
        index = int(out_data[i, j])
        cuda.atomic.add(local_hist, index, 1)
    
    cuda.syncthreads()
    
    if tid < 256:
        cuda.atomic.add(h2, tid, local_hist[tid])


def main():
    # Load image
    image = cv2.imread("./input/trees_big.jpg", cv2.IMREAD_GRAYSCALE)
    if image is None:
        print("Failed to load ./input/trees_big.jpg")
        return -1

    rows, cols = image.shape
    img_data = image.astype(np.uint8)

    # CUDA-specific: Define block and grid dimensions
    threads_per_block = (16, 16)  # 256 threads per block
    blocks_per_grid_x = (cols + threads_per_block[0] - 1) // threads_per_block[0]
    blocks_per_grid_y = (rows + threads_per_block[1] - 1) // threads_per_block[1]
    blocks_per_grid = (blocks_per_grid_x, blocks_per_grid_y)

    myMat1 = None

    for iter in range(15):

        with nvtx.annotate(f"iteration_{iter}"):

            arr = np.zeros(256, dtype=np.int32)
            arr2 = np.zeros(256, dtype=np.float32)
            arr3 = np.zeros(256, dtype=np.float32)
            h2 = np.zeros(256, dtype=np.int32)

            # ------------------------ histogram_original ------------------------
            with nvtx.annotate("histogram_original"):

                # CUDA-specific: Explicit device allocation with cuda.to_device
                d_img_data = cuda.to_device(img_data)
                d_arr = cuda.to_device(arr)

                # CUDA-specific: Kernel launch with [blocks, threads]
                histogram_kernel[blocks_per_grid, threads_per_block](d_img_data, d_arr, rows, cols)
                
                # CUDA-specific: Explicit synchronization
                cuda.synchronize()

                # CUDA-specific: Explicit copy back with copy_to_host()
                arr = d_arr.copy_to_host()

            # ------------------------ find_max_original_hist --------------------
            with nvtx.annotate("find_max_original_hist"):

                maxVal = 0
                for i in range(256):
                    if arr[i] > maxVal:
                        maxVal = arr[i]

            # ------------------------ pmf_compute -------------------------------
            with nvtx.annotate("pmf_compute"):

                total = float(rows * cols)
                for i in range(256):
                    arr2[i] = float(arr[i]) / total

            # ------------------------ cdf_compute -------------------------------
            with nvtx.annotate("cdf_compute"):

                arr3[0] = arr2[0]
                for i in range(1, 256):
                    arr3[i] = arr2[i] + arr3[i - 1]

            # ------------------------ apply_equalization -----------------------
            with nvtx.annotate("apply_equalization"):

                # CUDA-specific: Device array creation without host copy
                d_arr3 = cuda.to_device(arr3)
                d_out_data = cuda.device_array((rows, cols), dtype=np.uint8)

                apply_equalization_kernel[blocks_per_grid, threads_per_block](d_img_data, d_out_data, d_arr3, rows, cols)
                cuda.synchronize()

                out_data = d_out_data.copy_to_host()
                
                # CUDA-specific: Explicit memory cleanup
                del d_arr3

            # ------------------------ histogram_equalized -----------------------
            with nvtx.annotate("histogram_equalized"):

                d_out_data = cuda.to_device(out_data)
                d_h2 = cuda.to_device(h2)

                histogram_equalized_kernel[blocks_per_grid, threads_per_block](d_out_data, d_h2, rows, cols)
                cuda.synchronize()

                h2 = d_h2.copy_to_host()
                
                # CUDA-specific: Explicit memory cleanup
                del d_out_data
                del d_img_data
                del d_h2

            # ------------------------ find_max_equalized_hist -------------------
            with nvtx.annotate("find_max_equalized_hist"):

                maxH2 = 0
                for i in range(256):
                    if h2[i] > maxH2:
                        maxH2 = h2[i]

            myMat1 = out_data

    cv2.imwrite("./output/equalized_cuda_python.jpg", myMat1)
    print("Output saved to ./output/equalized_cuda_python.jpg")

    return 0


if __name__ == "__main__":
    main()